Make goals use C++20 coroutines

L-as opened the pull request:
Part of building refactor WG work. I don't necessarily expect to merge the code in the current state, but it's not really a draft since I think it works. There might be bugs hiding somewhere though. OTOH bugs we haven't discovered yet might have been removed by this PR.

# Motivation

Goals before this PR do coroutine-like things explicitly
using `state` and setting function pointers.
Everytime `work()` is called, the `state` function pointer is called.
The control flow resulting from this design is not very obvious (for me personally at least).
One consequence of this design is that any state that has to be maintained between
different `state`s has to be in the goal class, such that all variables that could
ever be used are all fields of the class.

It is because of this not clear what fields are used when,
and more importantly, which are initialized in which states.

It is further complicated by the fact that if you do not set the `state`,
it implicitly loops.

It is not a "nice" way of writing the goal logic.

C++20 however has coroutines, which can do much of what we do manually, automatically.

This PR is an experiment in that direction, seeing how goal logic would look
using coroutines.

# Context

Reference for learning about coroutines:
- https://lewissbaker.github.io/
- https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/coroutines-c++20/
- https://www.scs.stanford.edu/~dm/blog/c++-coroutines.html

My own as short as possible explanation:
Coroutines are functions that use co_await/co_return.
When you run the coroutine function, a coroutine handle is created,
alongside its user-defined promise type.
There are suspension points at the beginning, at every `co_await`,
and at the final (possibly implicit) `co_return`.
Once suspended, you can resume the coroutine by doing `coroutine_handle.resume()`.
Suspension points are implemented by passing a struct to the compiler
that implements `await_suspend`.
`await_suspend` can either say "cancel suspension", in which case execution resumes,
"suspend", in which case control is passed back to the caller of `coroutine_handle.resume()`
or the place where the coroutine function is initially executed in the case of the initial
suspension, or `await_suspend` can specify _another_ coroutine to jump to, which is
how tail calls are implemented.

---

I've tried to roughly make every state into a coroutine.
Where before you'd do `state = &f; return`, you now do `co_await SuspendGoal{}; co_return f()`.
However, you can also do `co_await f()`, i.e. a "coroutine" call!

Places where a suspension would happen implicitly (i.e. `return`-ing from the function),
are now explicit by doing `co_await SuspendGoal{}`.

Each goal at all times keeps track of the `top_co`, i.e. the coroutine last (possibly now) executed.
When we `co_return` another coroutine, i.e. do a "coroutine" tail call, we set the `top_co`,
and switch to it. Our current coroutine is destroyed.

When we `co_await` another coroutine, we _swap_ `top_co`, and set the previous `top_co`,
i.e. the coroutine being currently executed, as the _continuation_ of the coroutine to be called.

Once a coroutine has finished executing (at `co_return`), we check whether the goal is still busy
(is status `ecBusy`?), and if so, we jump to the continuation at the final suspension point.
If not, we suspend the coroutine and ignore the continuation, since the goal has finished.
The goal should be destroyed by someone else after this point.

Coroutines are wrapped in `Co` to ensure automatic destruction.

Currently, there is the limitation that our coroutines (`Co`) can not return a value,
but this should be trivial to implement using templates, or you could just pass in a pointer
to the coroutine and "return" the value that way.
It has however not been necessary yet.

I've rewritten most of the code mechanically.
`DrvOutputSubstitionGoal` I've used the most effort on, and it's an example
of how the logic can be heavily simplified.
The intermediate variables no longer have to be fields of the class and aren't in the
header anymore.

### Alternatives

You could have heavily simplified the code without coroutines too.
You could have made `state` into a function pointer with state,
i.e. a class, to improve the legibility of the control flow,
and also prevent implicit looping by making the state _return_ the
next state rather than setting a field that already has a value.

You could also have reimplemented the code in another language entirely.

I deemed both of these suboptimal. We're already using C++, so why not use
all of the features. Coroutines are mostly well supported by all major compilers
AFAICT, albeit GCC didn't support a feature that would reduce complexity by a small
amount (getting the `this` of coroutine methods in `promise_type` constructor).

There are _obviously_ better languages than C++, but deciding on a new language
is a social problem as well as a technical one. There also isn't a language
that is better than C++(20) on all fronts (e.g. bootstrappability, multiple implementations,
has a specification).

I feel that we can heavily improve the quality of the Nix codebase without switching from C++.

We are probably better off having multiple implementations of Nix anyway in my own humble opinion.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

L-as commented:
@Ericson2314

L-as commented:
@Ericson2314 @rhelmot

L-as commented:
I probably need to add more comments.

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
+        }
+    };;
+    struct promise_type {
+        // Either this is who called us, or it is who we will tail-call.
+        // It is what we "jump" to once we are done.
```
```suggestion
        /**
         * Either this is who called us, or it is who we will tail-call.
         * It is what we "jump" to once we are done.
         */
```

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
+
+        Co&& await_transform(Co&& co) { return static_cast<Co&&>(co); }
+        std::suspend_always await_transform(SuspendGoal) { return {}; };
+    };
```
```suggestion
    };

```

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
+        bool await_ready() { return false; };
+        std::coroutine_handle<> await_suspend(handle_type handle);
+        void await_resume() {};
+    };
```
```suggestion
    };

```

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
+        std::coroutine_handle<> await_suspend(handle_type handle);
+        void await_resume() {};
+    };
+    // Used on initial suspend, doesn't do anything useful,
+    // but asserts that everything has been set correctly.
```
```suggestion
    /**
     * Used on initial suspend, doesn't do anything useful,
     * but asserts that everything has been set correctly.
     */
```

Ericson2314 commented on src/libstore/build/goal.cc:
```diff
@@ -3,6 +3,105 @@
+    auto handle = handle_type::from_promise(*this);
+    return Co{handle};
+};
+// Here we execute our continuation, by passing it back to the caller.
+// C++ compiler will create code that takes that and executes it promptly.
+// `h` is the handle for the coroutine that is finishing execution,
+// thus it must be destroyed.
```
Move to header and make doxygen comment

Ericson2314 commented on src/libstore/build/goal.cc:
```diff
@@ -3,6 +3,105 @@
+    }
+}
+
+// When "returning" another coroutine, what happens is that
+// we set it as our own continuation, thus once the final suspend
+// happens, we transfer control to it.
+// The original continuation we had is set as the continuation
+// of the coroutine passed in.
+// `final_suspend` is called after this, and `final_awaiter` will pass control off to `continuation`.
+// However, we also have to transfer the ownership of `next`, since it's an rvalue,
+// the handle to which is on our stack.
+// We thus give it to our previous continuation.
```
Move to header and make doxygen comment

Ericson2314 commented on src/libstore/build/goal.cc:
```diff
@@ -3,6 +3,105 @@
+    continuation->handle.promise().continuation = std::move(old_continuation);
+}
+
+// When we `co_await` another `Co`-returning coroutine,
+// we tell the caller of `caller.resume()` to switch to our coroutine (`handle`).
+// To make sure we return to the original coroutine, we set it as the continuation of our
+// coroutine. In `final_awaiter` we check if it's set and if so we return to it.
+//
+// To explain in more understandable terms:
+// When we `co_await Co_returning_function()`, this function is called on the resultant Co of
+// the _called_ function, and C++ automatically passes the caller in.
+// We don't use this caller, because we make use of the invariant that top_co == caller.
```
Move to header and make doxygen comment

Ericson2314 commented on src/libstore/build/drv-output-substitution-goal.cc:
```diff
@@ -14,146 +14,133 @@ DrvOutputSubstitutionGoal::DrvOutputSubstitutionGoal(
+
+        worker.childTerminated(this);
+
+        /*
```
```suggestion
        /**
```
so Doxygen could see it

Ericson2314 commented on src/libstore/build/derivation-goal.cc:
```diff
@@ -107,13 +113,7 @@ void DerivationGoal::killChild()
-void DerivationGoal::work()
-{
-    (this->*state)();
+    auto _ = done(BuildResult::TimedOut, {}, std::move(ex));
```
what is being returned?

L-as replied:
From what I can tell this shouldn't actually work because `amDone` is a coroutine and we never execute it.
The point was that `amDone` wouldn't be a coroutine but would return something you would await that never resumes,
i.e. you are in a coroutine, call `amDone`, you `co_await` it, and that would be the final suspension.

I will fix this, but I don't know why the tests pass. In my mind they shouldn't.

L-as replied:
In the end I fixed the bug I talked about above and this line is as it was before, except I added [[maybe_unused]] to silence the warning.

There is probably a better way of doing this, by splitting `done` into e.g. a function that returns `Done` and one that doesn't return `Done`, but I imagine the code will change heavily anyway in my next PR, so I refrained.

So to answer you question for what it does _now_:

`amDone` (called by `done`) does what it did before, but returns a `Done`, which has `[[nodiscard]]`.
This is done such that in a coroutine when you do this, it's clear that you should `co_return` it immediately (such that the coroutine ends, which is most often the behavior you want).
It doesn't do anything though, it's just a way of making the user do `co_return` and causing `final_suspend` to happen.
If you're not in a coroutine in the first place, obviously this makes no sense in the first place, hence we ignore the `Done` value.

Ericson2314 replied:
Thanks! That makes sense.

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
+     * Suspend our goal and wait until we get work()-ed again.
+     */
+    struct SuspendGoal {};
+    struct Return {};
+    struct promise_type;
```
Should have doxygens

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
+    struct Return {};
+    struct promise_type;
+    using handle_type = std::coroutine_handle<promise_type>;
+    // FIXME: Allocate explicitly on stack since HALO thing doesn't really work,
+    // specifically, there's no way to uphold the requirements when trying to do
+    // tail-calls without using a trampoline AFAICT.
+    // NOTES:
+    // These are good resources for understanding how coroutines work:
+    // https://lewissbaker.github.io/
+    // https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/coroutines-c++20/
+    // https://www.scs.stanford.edu/~dm/blog/c++-coroutines.html
+    struct [[nodiscard]] Co {
```
Convert to doxygen, can use [`@todo`](https://www.doxygen.nl/manual/commands.html#cmdtodo) and [`@note`](https://www.doxygen.nl/manual/commands.html#cmdnote)

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -124,15 +211,18 @@ public:
-    { }
+        : worker(worker), top_co(init_wrapper())
+    {
+        // assert(!top_co->handle.promise().goal);
```
restore?

Ericson2314 commented on src/libstore/build/substitution-goal.cc:
```diff
@@ -102,29 +94,42 @@ void PathSubstitutionGoal::tryNext()
+        co_return tryNext();
     }
 
+    // Horrible, horrible code.
+    // Needed because we can't `co_*` inside a catch-clause.
+    // `std::variant` would be cleaner perhaps.
```
I image you can soon convert this stuff into a simpler for loop like you did with the other substitution goal?

Ericson2314 commented on src/libstore/build/goal.hh:
```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
+    // https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/coroutines-c++20/
+    // https://www.scs.stanford.edu/~dm/blog/c++-coroutines.html
+    struct [[nodiscard]] Co {
+        handle_type handle;
+        explicit Co(handle_type handle) : handle(handle) {};
+        Co(const Co&) = delete;
+        Co &operator=(const Co&) = delete;
+        void operator=(Co&&);
+        Co(Co&& rhs);
+        ~Co();
+
+        bool await_ready() { return false; };
+        std::coroutine_handle<> await_suspend(handle_type handle);
+        void await_resume() {};
```
The ones that don't have descriptions in the file (which should be moved here) should gain new doxygens

Ericson2314 commented on src/libstore/build/substitution-goal.cc:
```diff
@@ -12,20 +13,18 @@ PathSubstitutionGoal::PathSubstitutionGoal(const StorePath & storePath, Worker &
     maintainExpectedSubstitutions = std::make_unique<MaintainCount<uint64_t>>(worker.expectedSubstitutions);
 }
 
-
```
stray deleted line btw (not a big deal)

L-as commented:
Depends on #11073

L-as commented:
I tested this with memory sanitizer and address sanitizer and don't think I saw any new errors.

It seems to work? I rebased on master and removed the commits from #11073.

Ericson2314 reviewed (approved):
Let's go!
